### ISSUE TITLE
postInstallation fails on step "Creating a data snapshot"

### ISSUE THREAD

**Issue by viruz82** [bitnami/bitnami-docker-mariadb#138]
On bitnami mariadb 10.1
We use mesos/marathon, Debian 8.8.

On the master side we use a german language database (e.g. umlauts) which has a size of ~ 318 MB.
When I start the slave it hangs on the step 
`mariadb INFO  ==> Creating a data snapshot...`
and some seconds later it fails with:
```
Error executing 'postInstallation': "toString()" failed
W1207 09:52:43.860947 29369 logging.cpp:91] RAW: Received signal SIGTERM from process 2348 of user 0; exiting
```

When I set up a new nearly clear database everything works just fine.

**Comment by chipironcin** [comment]
Hello @viruz82 
It may be due to the character set.
The default character set is `latin` and I see some suggests to use `utf-8` for supporting special characters like umlauts.
https://stackoverflow.com/questions/5526169/what-is-the-best-mysql-collation-for-german-language

Please refer to this guide:
https://mariadb.com/kb/en/library/setting-character-sets-and-collations/

**Comment by viruz82** [comment]
Hello @chipironcin ,

thanks for your hint.

We used `utf8_general_ci`. Just tested to switch to `utf8_unicode_ci` but the error accured again.
I just saw that the databse is not 318 MB but 560 MB. Could it have something to do with the database size?

**Comment by tompizmor** [comment]
Hello @viruz82,

Could you please set the following env var and reproduce the issue again? Paste the error trace here.

```
NAMI_LOG_LEVEL=trace8
```
> Could it have something to do with the database size?

It is possible yes. In order to check it, you could try using a smaller database with the same kind of characters and make a test.

**Comment by viruz82** [comment]
Hello @tompizmor ,

following the log:

`Registered docker executor on srv-mesos-slave6.xxx.xxx.de
Starting task test_lms-mariadb-slave.4aed4baa-e0bf-11e7-b337-005056a128d3
mariadb TRACE [normalize] File normalization: conf -> /opt/bitnami/mariadb/conf
mariadb TRACE [normalize] File normalization: bin -> /opt/bitnami/mariadb/bin
mariadb TRACE [normalize] File normalization: data -> /opt/bitnami/mariadb/data
mariadb TRACE [normalize] File normalization: logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [normalize] File normalization: lib -> /opt/bitnami/mariadb/lib
mariadb TRACE [normalize] File normalization: conf -> /opt/bitnami/mariadb/conf
mariadb TRACE [normalize] File normalization: bin -> /opt/bitnami/mariadb/bin
mariadb TRACE [normalize] File normalization: data -> /opt/bitnami/mariadb/data
mariadb TRACE [normalize] File normalization: logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [normalize] File normalization: lib -> /opt/bitnami/mariadb/lib
mariadb TRACE [normalize] File normalization: bin -> /opt/bitnami/mariadb/bin
mariadb TRACE [normalize] File normalization: conf -> /opt/bitnami/mariadb/conf
mariadb TRACE [normalize] File normalization: bin -> /opt/bitnami/mariadb/bin
mariadb TRACE [normalize] File normalization: data -> /opt/bitnami/mariadb/data
mariadb TRACE [normalize] File normalization: logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [normalize] File normalization: lib -> /opt/bitnami/mariadb/lib
mariadb TRACE [normalize] File normalization: conf -> /opt/bitnami/mariadb/conf
mariadb TRACE [normalize] File normalization: bin -> /opt/bitnami/mariadb/bin
mariadb TRACE [normalize] File normalization: data -> /opt/bitnami/mariadb/data
mariadb TRACE [normalize] File normalization: logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [normalize] File normalization: lib -> /opt/bitnami/mariadb/lib
mariadb TRACE [normalize] File normalization: conf -> /opt/bitnami/mariadb/conf
mariadb TRACE [normalize] File normalization: bin -> /opt/bitnami/mariadb/bin
mariadb TRACE [normalize] File normalization: data -> /opt/bitnami/mariadb/data
mariadb TRACE [normalize] File normalization: logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [normalize] File normalization: lib -> /opt/bitnami/mariadb/lib
mariadb TRACE [normalize] File normalization: bin -> /opt/bitnami/mariadb/bin
nami    INFO  Initializing mariadb
mariadb TRACE [normalize] File normalization: conf -> /opt/bitnami/mariadb/conf
mariadb TRACE [normalize] File normalization: data -> /opt/bitnami/mariadb/data
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [mkdir] File normalization: /opt/bitnami/mariadb/conf -> /opt/bitnami/mariadb/conf
mariadb TRACE [mkdir] File normalization: /opt/bitnami/mariadb/data -> /opt/bitnami/mariadb/data
mariadb TRACE [mkdir] File normalization: /opt/bitnami/mariadb/tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [mkdir] File normalization: /opt/bitnami/mariadb/logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [exists] File normalization: /bitnami/mariadb/.initialized -> /bitnami/mariadb/.initialized
mariadb INFO  ==> Cleaning data dir...
mariadb TRACE [normalize] File normalization: data -> /opt/bitnami/mariadb/data
mariadb TRACE [delete] File normalization: /opt/bitnami/mariadb/data/* -> /opt/bitnami/mariadb/data/*
mariadb INFO  ==> Configuring permissions...
mariadb INFO  ==> Validating inputs...
mariadb INFO  ==> Initializing database...
mariadb TRACE [normalize] File normalization: conf -> /opt/bitnami/mariadb/conf
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/conf/my.cnf -> /opt/bitnami/mariadb/conf/my.cnf
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysql.sock -> /opt/bitnami/mariadb/tmp/mysql.sock
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/logs/mysqld.log -> /opt/bitnami/mariadb/logs/mysqld.log
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysql.sock -> /opt/bitnami/mariadb/tmp/mysql.sock
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysql.sock -> /opt/bitnami/mariadb/tmp/mysql.sock
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: conf -> /opt/bitnami/mariadb/conf
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/conf/my.cnf -> /opt/bitnami/mariadb/conf/my.cnf
mariadb TRACE [runProgram] Executing: /opt/bitnami/mariadb/bin/mysql_install_db --defaults-file=/opt/bitnami/mariadb/conf/my.cnf
mariadb TRACE [runProgram] Executing internal command: /opt/bitnami/mariadb/bin/mysql_install_db ["--defaults-file=/opt/bitnami/mariadb/conf/my.cnf"]
mariadb TRACE [runProgram] RESULT: {"code":0,"stderr":"2017-12-14 11:08:07 140438733137792 [Note] /opt/bitnami/mariadb/sbin/mysqld (mysqld 10.1.29-MariaDB) starting as process 51 ...\n2017-12-14 11:08:13 140401745627008 [Note] /opt/bitnami/mariadb/sbin/mysqld (mysqld 10.1.29-MariaDB) starting as process 80 ...\n2017-12-14 11:08:16 140671804962688 [Note] /opt/bitnami/mariadb/sbin/mysqld (mysqld 10.1.29-MariaDB) starting as process 109 ...\n","stdout":"Installing MariaDB/MySQL system tables in './data' ...\nOK\nFilling help tables...\nOK\nCreating OpenGIS required SP-s...\nOK\n\nTo start mysqld at boot time you have to copy\nsupport-files/mysql.server to the right place for your system\n\nPLEASE REMEMBER TO SET A PASSWORD FOR THE MariaDB root USER !\nTo do so, start the server, then issue the following commands:\n\n'/opt/bitnami/mariadb/bin/mysqladmin' -u root password 'new-password'\n'/opt/bitnami/mariadb/bin/mysqladmin' -u root -h 172.18.0.28 password 'new-password'\n\nAlternatively you can run:\n'/opt/bitnami/mariadb/bin/mysql_secure_installation'\n\nwhich will also give you the option of removing the test\ndatabases and anonymous user created by default.  This is\nstrongly recommended for production servers.\n\nSee the MariaDB Knowledgebase at http://mariadb.com/kb or the\nMySQL manual for more instructions.\n\nYou can start the MariaDB daemon with:\ncd '/opt/bitnami/mariadb' ; /opt/bitnami/mariadb/bin/mysqld_safe --datadir='./data'\n\nYou can test the MariaDB daemon with mysql-test-run.pl\ncd '/opt/bitnami/mariadb/mysql-test' ; perl mysql-test-run.pl\n\nPlease report any problems at http://mariadb.org/jira\n\nThe latest information about MariaDB is available at http://mariadb.org/.\nYou can find additional information about the MySQL part at:\nhttp://dev.mysql.com\nConsider joining MariaDB's strong and vibrant community:\nhttps://mariadb.org/get-involved/\n\n"}
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/logs/mysqld.log -> /opt/bitnami/mariadb/logs/mysqld.log
mariadb TRACE [normalize] File normalization: logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/logs/mysqld.log -> /opt/bitnami/mariadb/logs/mysqld.log
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/logs/mysqld.log -> /opt/bitnami/mariadb/logs/mysqld.log
mariadb TRACE [normalize] File normalization: logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/logs/mysqld.log -> /opt/bitnami/mariadb/logs/mysqld.log
mariadb TRACE [exists] File normalization: /opt/bitnami/mariadb/logs -> /opt/bitnami/mariadb/logs
mariadb TRACE [exists] File normalization: /tmp -> /tmp
mariadb TRACE [exists] File normalization: /tmp -> /tmp
mariadb TRACE [normalize] File normalization: conf -> /opt/bitnami/mariadb/conf
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/conf/my.cnf -> /opt/bitnami/mariadb/conf/my.cnf
mariadb TRACE [runProgram] Executing: null ./bin/mysqld_safe --defaults-file=/opt/bitnami/mariadb/conf/my.cnf --binlog-format=ROW --log-bin=mysql-bin --sync-binlog=1 --server-id=104 --relay-log=mysql-relay-bin --log-slave-updates=1 --read-only=1
mariadb TRACE [runProgram] Executing internal command: '/bin/sh' ["-c","./bin/mysqld_safe --defaults-file=/opt/bitnami/mariadb/conf/my.cnf --binlog-format=ROW --log-bin=mysql-bin --sync-binlog=1 --server-id=104 --relay-log=mysql-relay-bin --log-slave-updates=1 --read-only=1 &"]
mariadb TRACE [runProgram] RESULT: {"code":0,"stderr":"","stdout":""}
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE [normalize] File normalization: tmp -> /opt/bitnami/mariadb/tmp
mariadb TRACE [normalize] File normalization: /opt/bitnami/mariadb/tmp/mysqld.pid -> /opt/bitnami/mariadb/tmp/mysqld.pid
mariadb TRACE code: 0
mariadb TRACE stdout:

mariadb TRACE stderr:

mariadb INFO  ==> Configuring replication mode slave...
mariadb INFO  ==> Setting up database slave...
mariadb INFO  ==> Checking if replication master is ready to accept connection ...
INFO  Trying to connect to MySQL server
INFO  Found MySQL server listening at lms-mariadb.backend.xxx.xxx.de:3307
INFO  MySQL server listening and working at lms-mariadb.backend.xxx.xxx.de:3307
mariadb INFO  ==> Setting the master configuration...
mariadb INFO  ==> Creating a data snapshot...
mariadb TRACE [runProgram] Executing: /opt/bitnami/mariadb/bin/mysqldump -uroot,-hlms-mariadb.backend.xxx.xxx.de,-P3307,--all-databases,--skip-lock-tables,--single-transaction,--flush-logs,--hex-blob,--master-data,--apply-slave-statements,--comments=false,-pmasTeR23
mariadb TRACE [runProgram] Executing internal command: /opt/bitnami/mariadb/bin/mysqldump ["-uroot","-hlms-mariadb.backend.xxx.xxx.de","-P3307","--all-databases","--skip-lock-tables","--single-transaction","--flush-logs","--hex-blob","--master-data","--apply-slave-statements","--comments=false","-pmasTeR23"]
nami    TRACE Error: Error executing 'postInstallation': "toString()" failed
    at Buffer.toString (buffer.js:528:11)
    at runProgram (/opt/bitnami/nami/node_modules/nami-utils/lib/os/run-program.js:197:37)
    at Object.runProgram (/opt/bitnami/nami/node_modules/nami-utils/lib/os/index.js:63:12)
    at Object.$app.helpers.configureReplication (/.nami/components/com.bitnami.mariadb/helpers.js:227:11)
    at Service.$app.postInstallation (/.nami/components/com.bitnami.mariadb/main.js:94:20)
    at Service.runStep (/opt/bitnami/nami/node_modules/nami-core/lib/components/component.js:159:12)
    at _.each.step (/opt/bitnami/nami/node_modules/nami-core/lib/components/component.js:236:32)
    at ot (/opt/bitnami/nami/node_modules/lodash/index.js:12:507)
    at Function.<anonymous> (/opt/bitnami/nami/node_modules/lodash/index.js:33:421)
    at Service.install (/opt/bitnami/nami/node_modules/nami-core/lib/components/component.js:236:7)`

I will test it soon with a smaller database and the same kind of characters.

Cheers

**Comment by tompizmor** [comment]
Hello @viruz82,

Did you find the time to test a smaller database? 

I've been doing some tests on my own and I think the issue could be that the output of the `mysqldump` command is too verbose and is making the toString() method to fail.

Would you mind trying the following to check if it works for you?

1. Edit the _docker-compose.yml_ file to override the startup command for the mariadb-slave service. Just append the command property. As the mariadb runs as non-root by default you need to specify the `root` user to be able to edit a file later.

```
...
  mariadb-slave:
    image: 'bitnami/mariadb:latest'
    user: root
    command: tail -f /dev/null
...
```

2. Exec into the container

3. Install a text editor (vim or emacs)

```
$ install_packages vim emacs24-nox
```

4. Edit the file _/.nami/components/com.bitnami.mariadb/helpers.js_ and change the line 227 from 

```
$os.runProgram('bin/mysqldump', commandArguments);
```
to this

```
$os.runProgram('bin/mysqldump', commandArguments, {ignoreStdStreams: true});
```

5. Initialize the mariadb slave with these commands.

```
$ . /opt/bitnami/base/functions
$ . /opt/bitnami/base/helpers
$   nami_initialize mariadb
```

Finally, check if the data snapshot is successfully done.

**Comment by viruz82** [comment]
Hello @tompizmor ,

I have just tested and your fix was successful. The snapshot is there and the container runs without errors, yeah. :+1: 

Thank you very much!

Was this just a temporary fix or do you plan to fix it in your repository?

Cheers

**Comment by tompizmor** [comment]
I am glad it was helpful for you @viruz82!

It was just a temporary fix. I will release a new revision of the mariadb docker image including the fix. I will post here once it is available in order to close the issue.

**Comment by tompizmor** [comment]
@viruz82 I just released mariadb `10.1.29-r2` and mysql `5.7.20-r2` including the fix.
